Add package wtxmgr to use walletdb interface

tuxcanfly opened the pull request:
**NOTE: This pull request requires #147**

Moved `txstore` to `legacy` and added package `wtxmgr` which provides `walletdb` support.

Legacy `txstore` will be imported as a part of the wallet setup as implemented in #147.

Working on docs, cleanup and rebase.

davecgh commented on wtxmgr/doc.go:
```diff
@@ -0,0 +1,102 @@
+// outputs spendable with wallet keys and transactions that are signed by
+// wallet keys in memory, handle spend tracking for newly-inserted
+// transactions, report the spendable balance from each unspent
+// transaction output, and finally to provide a means to serialize the
```
This section is no longer true.

davecgh commented on wtxmgr/error.go:
```diff
@@ -0,0 +1,92 @@
+}
+
+// TxStoreError provides a single type for errors that can happen during tx
+// store operation. It is similar to waddrmgr.ManagerError.
```
Imo, no need to call out it's similar to waddrmgr.ManagerError here since the packages are largely unrelated.

davecgh commented on wtxmgr/json.go:
```diff
@@ -0,0 +1,198 @@
+	CreditImmature
+)
+
+// category returns the category of the credit.  The passed block chain height is
```
Category

davecgh commented on wtxmgr/json.go:
```diff
@@ -0,0 +1,198 @@
+	return c.category(chainHeight)
+}
+
+func (c *Credit) category(chainHeight int32) CreditCategory {
```
No comment

davecgh commented on wtxmgr/json.go:
```diff
@@ -0,0 +1,198 @@
+	return d.toJSON(account, chainHeight, net)
+}
+
+func (d Debits) toJSON(account string, chainHeight int32,
```
As noted else, please define funcs before they're called for consistency.

davecgh commented on wtxmgr/json.go:
```diff
@@ -0,0 +1,198 @@
+	return c.toJSON(account, chainHeight, net)
+}
+
+func (c Credit) toJSON(account string, chainHeight int32,
```
Same.

davecgh commented on wtxmgr/notifications.go:
```diff
@@ -0,0 +1,145 @@
+// ErrDuplicateListen is returned for any attempts to listen for the same
+// notification more than once.  If callers must pass along a notifiation to
+// multiple places, they must broadcast it themself.
+var ErrDuplicateListen = errors.New("duplicate listen")
```
All errors coming out of the manager should be of type TxManagerError.

davecgh commented:
There is a major lack of comments.  Tons of unexported functions without comments, details not explained, etc.

tuxcanfly commented:
@davecgh Sorry, let me get those and also incorporate suggestions from the previous PR and get back to you before reviewing this.

davecgh commented:
@tuxcanfly No rush on this.  It's going to take a while to get through it all and I still have more to review on the other PR too.  Just was taking a quick perusal through this.

tuxcanfly commented:
OK, thanks. I have already addressed most of the items in the other PR, so shouldn't be long before getting back to this.

tuxcanfly commented:
Rebased on top of `btcsuite` migration. Added docs for all newly introduced changes. Will be cross checking for addressable items from previous reviews, but should generally be good for review now.

davecgh commented:
One of the things we'll need before this can go in is a new RPC to force the txstore to be recreated.  Perhaps something like "recalculatebalances".  Previously, if anything got out of sync, we could simply delete the separate file to force this to happen, but that will no longer be possible once it's part of the overall wallet database.

This becomes particularly important when BIP0032 keys are being used since a user could easily generate the next keys in sequence from something external and receive funds to them.  The wallet would obviously not yet know about these.  Subsequent calls to getnewaddress would, of course, make them known to wallet, but at that point it won't be rescanning those earlier blocks where the funds were received.

jrick commented:
We will want this in before merging the walletdb integration.  Can you rebase on top of the intwaddrmgr branch?

Also if possible, can you switch the PR from master to intwaddrmgr so it only includes new commits not already on intwaddrmgr? (I don't know if github allows this)

tuxcanfly commented:
Rebased on top of intwaddrmgr. Looks like we'll need to create a new PR with the base branch as intwaddrmgr instead of master.

jrick commented:
Okay, it's a pain but I think that's the best solution.  You can close this one and reference it from the new one.

tuxcanfly commented:
Done, closing in favor of #190
